assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js

```diff
@@ -0,0 +1,162 @@
+      getMedia({ searchTerm, pagingNum: p })
+        .then(({ data }) => {
+          setMedia({
+            media: data,
+          });
+        })
+        .catch(fetchMediaError);
+    },
+    [setMedia, fetchMediaError, fetchMediaStart, getMedia, searchTerm]
+  );
+
+  const uploadMediaFromLibrary = useCallback(
+    async (files) => {
+      try {
+        const localMedia = await Promise.all(
+          files.map(getResourceFromLocalFile).reverse()
+        );
+        const filesUploading = files.map((file) => uploadFile(file));
+        setMedia({
+          media: [...localMedia, ...media],
+        });
+        await Promise.all(filesUploading);
+
+        // To avoid race conditions updating media library, a new request is necessary
+        resetMedia({ pagingNum });
+      } catch (e) {
+        setMedia({ media });
+        fetchMediaError(e);
+      }
+    },
+    [resetMedia, setMedia, fetchMediaError, uploadFile, media, pagingNum]
+  );
+
+  const uploadMediaFromWorkspace = useCallback(
+    async (files, { insertElement, updateElementById, deleteElementById }) => {
+      try {
+        const filesOnCanvas = await Promise.all(
+          files
+            .map(async (file) => {
+              const resource = await getResourceFromLocalFile(file);
+              const element = insertElement(resource.type, {
+                resource: getResourceFromAttachment(resource),
+              });
+
+              return {
+                resource,
+                element,
+                file,
+              };
+            })
+            .reverse()
+        );
+        const uploadFileOnCanvas = async ({ element, file }) => {
+          try {
+            const uploadedFile = await uploadFile(file);
+
+            const resource = getResourceFromUploadAPI(uploadedFile);
+
+            updateElementById({
+              elementId: element.elementId,
+              properties: {
+                resource,
+                type: element.resource.type,
+              },
+            });
+
+            return resource;
+          } catch (e) {
+            deleteElementById({ elementId: element.id });
+            setMedia({
+              media: media.filter(({ id }) => element.resource.id !== id),
+            });
+            fetchMediaError(e);
+            throw new Error('Error uploading a file.');
+          }
+        };
+
+        setMedia({
+          media: [...filesOnCanvas.map(({ resource }) => resource), ...media],
+        });
+
+        await Promise.all(filesOnCanvas.map(uploadFileOnCanvas));
+
+        // To avoid race conditions updating media library, a new request is necessary
+        resetMedia({ pagingNum });
+      } catch (e) {
+        setMedia({ media });
+        fetchMediaError(e);
+      }
+    },
+    [resetMedia, setMedia, fetchMediaError, uploadFile, media, pagingNum]
+  );
+
+  const uploadMedia = useCallback(
+    (files, { insertElement, updateElementById, deleteElementById } = {}) => {
```

Comment:
I double-checked and I believe I made an error about `useInsertElement()`. It looks like it's used in the library panel and thus available here. And the `updateElementById` and `deleteElementById` should definitely be available. I think it'd be better to resolve them up top then drag them as params in this case. Alternatively we could do a callback that looks like this:

```
uploadMedia(files, {onLocalFile, onUploadedFile, onUploadFailure})
```

Then we'd be pretty independent of the exact use. And we could leave it up to the `WorkspaceUploadTarget` to do the right thing - that would split code a little better, I think,.